File: shared/sdk-ts/src/event.ts
/**
 * Canonical event envelope on the NATS bus (ARCHITECTURE.md §6).
 * M0 skeleton: types + helpers are stable; the NATS transport lands in M1.
 */

export const ENVELOPE_VERSION = 1;

export interface Envelope<P = unknown> {
  id: string;            // UUID; consumers dedupe on this
  stream: string;        // JetStream stream name
  type: string;          // e.g. "task.finished"
  ts: string;            // ISO-8601 (UTC)
  trace_id?: string;     // OpenTelemetry trace id
  subject: string;       // NATS subject, e.g. "task:1234"
  payload: P;            // type-specific (see shared/proto/events.json)
  version: number;       // envelope schema version
}

export class EnvelopeError extends Error {}

/** Validate the version-independent invariants of an envelope. */
export function validateEnvelope(env: Envelope): void {
  if (!env.id) throw new EnvelopeError("envelope.id is required");
  if (!env.stream || !env.type || !env.subject) {
    throw new EnvelopeError("envelope stream/type/subject are required");
  }
  if (env.version !== ENVELOPE_VERSION) {
    throw new EnvelopeError(
      `envelope version mismatch: ${env.version} != ${ENVELOPE_VERSION}`,
    );
  }
  if (env.payload === undefined || env.payload === null) {
    throw new EnvelopeError("envelope.payload is required");
  }
}

/** Build a validated envelope with a generated id + timestamp. */
export function newEnvelope<P>(
  stream: string,
  type: string,
  subject: string,
  payload: P,
  trace_id?: string,
): Envelope<P> {
  const env: Envelope<P> = {
    id: randomUuid(),
    stream,
    type,
    subject,
    payload,
    ts: new Date().toISOString(),
    trace_id,
    version: ENVELOPE_VERSION,
  };
  validateEnvelope(env);
  return env;
}

function randomUuid(): string {
  const g = globalThis as { crypto?: { randomUUID?: () => string } };
  if (g.crypto?.randomUUID) return g.crypto.randomUUID(); // Node 16.7+/browsers
  // RFC 4122 v4 fallback for old runtimes.
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0;
    return (c === "x" ? r : (r & 0x3) | 0x8).toString(16);
  });
}

File: shared/sdk-go/event.go
// Package aisdlc provides the shared event-envelope type and a minimal bus
// abstraction for the Agentic SDLC Platform. See shared/proto/README.md.
//
// This is an M0 skeleton: the NATS JetStream transport and OpenTelemetry
// propagation land with the first real consumer (M1+). The Envelope type and
// the Idempotent handler helper are stable. The package is stdlib-only.
package aisdlc

import (
	"encoding/json"
	"errors"
	"fmt"
	"time"
)

// EnvelopeVersion is the canonical event-envelope schema version.
const EnvelopeVersion = 1

// Envelope is the canonical event wrapper published on the NATS bus
// (ARCHITECTURE.md §6). time.Time marshals to RFC 3339 by default.
type Envelope struct {
	ID      string          `json:"id"`                 // UUID; consumers dedupe on this
	Stream  string          `json:"stream"`             // JetStream stream name
	Type    string          `json:"type"`               // e.g. "task.finished"
	TS      time.Time       `json:"ts"`                 // when produced (UTC)
	TraceID string          `json:"trace_id,omitempty"` // OpenTelemetry trace id
	Subject string          `json:"subject"`            // NATS subject, e.g. "task:1234"
	Payload json.RawMessage `json:"payload"`            // type-specific; see events.json
	Version int             `json:"version"`            // envelope schema version
}

// Validate checks the envelope's version-independent invariants.
func (e Envelope) Validate() error {
	if e.ID == "" {
		return errors.New("aisdlc: envelope.id is required")
	}
	if e.Stream == "" || e.Type == "" || e.Subject == "" {
		return errors.New("aisdlc: envelope stream/type/subject are required")
	}
	if e.Version < 1 {
		return fmt.Errorf("aisdlc: envelope version mismatch: got %d, minimum is 1", e.Version)
	}
	if len(e.Payload) == 0 || string(e.Payload) == "null" {
		return errors.New("aisdlc: envelope.payload is required")
	}
	return nil
}
